controllers/iamauthpolicy_controller.go

```diff
@@ -29,34 +36,156 @@ func RegisterIAMAuthPolicyController(log gwlog.Logger, mgr ctrl.Manager) error {
 }
 
 func (c *IAMAuthPolicyController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	c.log.Infow("reconcile", "req", req)
-
-	policy := &anv1alpha1.IAMAuthPolicy{}
-	err := c.client.Get(ctx, req.NamespacedName, policy)
-	if !k8serr.IsNotFound(err) {
-		return ctrl.Result{}, err
+	k8sPolicy := &anv1alpha1.IAMAuthPolicy{}
+	err := c.client.Get(ctx, req.NamespacedName, k8sPolicy)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
+	c.log.Infow("reconcile", "req", req, "targetRef", k8sPolicy.Spec.TargetRef)
+
+	c.handleFinalizer(ctx, k8sPolicy)
 
-	switch policy.Spec.TargetRef.Kind {
+	isDelete := !k8sPolicy.DeletionTimestamp.IsZero()
+	kind := k8sPolicy.Spec.TargetRef.Kind
+	var reconcileFunc func(context.Context, *anv1alpha1.IAMAuthPolicy) (string, error)
+	switch kind {
 	case "Gateway":
-		err = c.reconcileGateway(ctx, policy)
-		break
-	case "HTTPRoute":
-	case "GRPCRoute":
-		err = c.reconcileRoute(ctx, policy)
-		break
-	default:
-		err = fmt.Errorf("unsupported targetRef type, req=%s, kind=%s",
-			req, policy.Spec.TargetRef.Kind)
+		if isDelete {
+			reconcileFunc = c.deleteGatewayPolicy
+		} else {
+			reconcileFunc = c.upsertGatewayPolicy
+		}
+	case "HTTPRoute", "GRPCRoute":
+		if isDelete {
+			reconcileFunc = c.deleteRoutePolicy
+		} else {
+			reconcileFunc = c.upsertRoutePolicy
+		}
+	}
+
+	latticeResourceId, err := reconcileFunc(ctx, k8sPolicy)
+	if err != nil {
+		if services.IsNotFoundError(err) {
+			c.log.Infof("reconcile error, retry in 30sec: %s", err)
+			return ctrl.Result{RequeueAfter: time.Second * 30}, nil
+		}
+		return ctrl.Result{}, err
 	}
+
+	k8sPolicy.Annotations["application-networking.k8s.aws/resourceId"] = latticeResourceId
+
+	err = c.client.Update(ctx, k8sPolicy)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
-	c.log.Infow("successfully reconciled", "req", req)
+	c.log.Infow("reconciled IAM policy",
+		"req", req,
+		"targetRef", k8sPolicy.Spec.TargetRef,
+		"latticeResorceId", latticeResourceId,
+		"isDeleted", isDelete,
+	)
 	return ctrl.Result{}, nil
 }
 
+func (c *IAMAuthPolicyController) handleFinalizer(ctx context.Context, k8sPolicy *anv1alpha1.IAMAuthPolicy) {
+	authPolicyFinalizer := "iamauthpolicy.k8s.aws/resources"
+	if k8sPolicy.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(k8sPolicy, authPolicyFinalizer) {
+			controllerutil.AddFinalizer(k8sPolicy, authPolicyFinalizer)
+		}
+	} else {
+		if controllerutil.ContainsFinalizer(k8sPolicy, authPolicyFinalizer) {
+			controllerutil.RemoveFinalizer(k8sPolicy, authPolicyFinalizer)
```

Comment:
When handling resouce deletion, we should do `RemoveFinalizer()` at very last, to avoid actual reconcile logic return err and k8s api server lose track of that IAMAuthPolicy resource?

For example, in our other controller:
https://github.com/aws/aws-application-networking-k8s/blob/658bb6e2cac0a8c5b36c6c0ac9a80a4d2487b412/controllers/route_controller.go#L195-L204